src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,974 @@
+            }
+
+            return (actionsBuilder.ToImmutableAndClear(), actionsBuilderAllOccurrences.ToImmutableAndClear());
+
+            // Local function to create a code action with more ease
+            MyCodeAction CreateNewCodeAction(string actionName, bool allOccurrences, SelectedCodeAction selectedCodeAction)
+            {
+                return new MyCodeAction(actionName, c => IntroduceParameterAsync(
+                    document, expression, methodSymbol, containingMethod, allOccurrences, selectedCodeAction, c));
+            }
+        }
+
+        /// <summary>
+        /// Introduces a new parameter and refactors all the call sites based on the selected code action.
+        /// </summary>
+        private async Task<Solution> IntroduceParameterAsync(Document originalDocument, TExpressionSyntax expression,
+            IMethodSymbol methodSymbol, SyntaxNode containingMethod, bool allOccurrences, SelectedCodeAction selectedCodeAction,
+            CancellationToken cancellationToken)
+        {
+            var parameterName = await GetNewParameterNameAsync(originalDocument, expression, cancellationToken).ConfigureAwait(false);
+
+            var methodCallSites = await FindCallSitesAsync(originalDocument, methodSymbol, cancellationToken).ConfigureAwait(false);
+
+            var modifiedSolution = originalDocument.Project.Solution;
+            var mappingDictionary = await MapExpressionToParametersAsync(originalDocument, expression, cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = originalDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            foreach (var (project, projectCallSites) in methodCallSites.GroupBy(kvp => kvp.Key.Project))
+            {
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var (document, invocations) in projectCallSites)
+                {
+                    var insertionIndex = GetInsertionIndex(compilation, methodSymbol, syntaxFacts, containingMethod);
+                    if (selectedCodeAction is SelectedCodeAction.Trampoline or SelectedCodeAction.Overload)
+                    {
+                        var newRoot = await ModifyDocumentInvocationsTrampolineOverloadAndIntroduceParameterAsync(compilation,
+                            document, originalDocument, invocations, mappingDictionary, methodSymbol, containingMethod,
+                            insertionIndex, allOccurrences, parameterName, expression, selectedCodeAction, cancellationToken).ConfigureAwait(false);
+                        modifiedSolution = modifiedSolution.WithDocumentSyntaxRoot(originalDocument.Id, newRoot);
+                    }
+                    else
+                    {
+                        var newRoot = await ModifyDocumentInvocationsAndIntroduceParameterAsync(compilation,
+                            originalDocument, document, mappingDictionary, containingMethod,
```

Comment:
the reversing of originalDocument/document here versus ModifyDocumentInvocationsTrampolineOverloadAndIntroduceParameterAsync will give me a brain embolism :)

Reply:
switched lol